core/server/master/src/test/java/alluxio/master/meta/MetaDailyBackupTest.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.meta;
+
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import alluxio.AlluxioURI;
+import alluxio.Configuration;
+import alluxio.ConfigurationRule;
+import alluxio.PropertyKey;
+import alluxio.master.BackupManager;
+import alluxio.underfs.UfsFileStatus;
+import alluxio.underfs.UfsStatus;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.util.CommonUtils;
+import alluxio.util.executor.ControllableScheduler;
+import alluxio.util.io.PathUtils;
+import alluxio.wire.BackupResponse;
+
+import com.google.common.collect.ImmutableMap;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import java.io.Closeable;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Tests the {@link MetaDailyBackup}.
+ */
+public class MetaDailyBackupTest {
+  private MetaMaster mMetaMaster;
+  private ControllableScheduler mScheduler;
+  private UnderFileSystem mUfs;
+  private Random mRandom;
+
+  @Test
+  public void test() throws Exception {
+    try (Closeable c =
+        new ConfigurationRule(ImmutableMap.of(PropertyKey.MASTER_DAILY_BACKUP_ENABLED, "true",
+        PropertyKey.MASTER_DAILY_BACKUP_FILES_RETAINED, "1")).toResource()) {
+      MetaDailyBackup dailyBackup = new MetaDailyBackup(mMetaMaster, mScheduler, mUfs);
+      dailyBackup.start();
+
+      // Trigger the runnable within 3 days
+      mScheduler.jumpToFuture(3, TimeUnit.DAYS);
+
+      verify(mMetaMaster, times(3)).backup(any());
+      verify(mUfs, times(3)).listStatus(any());
+      verify(mUfs, times(6)).deleteFile(any());
```

Comment:
Calling `backup` 3 times makes sense to me. Without knowing anything about the implementation, I would expect that backup will be executed 3 times in 3 days.

On the other hand, there's no way to know that we will call `listStatus` 3 times. Metadata backup could be implemented without calling `listStatus` at all, or it could call it once, or it could call it more than 3 times. It is an implementation detail that users of `MetaDailyBackup` don't need to know about, similar to how many times we use the `+` operation over the course of the test.

To test whether we deleted stale files, we should use real implementations instead of mocks, and directly check the backup directory to see whether stale files have been removed. That gives a lot more confidence that the feature is actually working, and the test won't break if the implementation changes.

Reply:
I think keeping `deleteFile` is ok, but can we use constants to show where the `6` comes from?

Reply:
Improved the `DailyMetadataTest` to include `fileToRetain, backUpFileNum, deleteFileNum`so  viewers can know how the number is calculated. Thanks